keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md

```diff
@@ -0,0 +1,832 @@
+{Resource: "apps.k8s.io/deployments", DeletionPriority: "10"}            
+...
+```
+0 would be the default DeletionPriority if not specified. And the resources which have deletion ordering concern would make sure to set DeletionPriority ahead of the resource it guards.
+
+Option 2: have the deletion priority bands defined instead of using numbers.
+
+To begin with, the deletion order bands would be introduced:
+- Workload Controllers
+- Workloads
+- Default
+- Policies
+
+Those deletion order bands will be deleted in sequence. E.g.
+
+```
+{Resource: "/pods", DeletionPriority: "workload"}
+{Resource: "networking.k8s.io/networkpolicies", DeletionPriority: "policies"}                     
+{Resource: "apps.k8s.io/dloyments", DeletionPriority: "workload controllers"} 
+...
+```
+
+After the deletion priority set, the namespace deletion process will honor the priority and delete the resources in sequence. 
+The resources with higher priority should always be deleted before the resources with lower priority.
+
+<References> Looks like we have examples for using int value for priority configurations existing in Kubernetes already.
+
+1.[Aggregator API priority] https://github.com/kubernetes/kubernetes/blob/659c437b267c4535d2855beee8abe5c121d58569/cmd/kube-apiserver/app/aggregator.go#L27-L59
+
+2. [API priority and fairness](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/#flowschema)
+
+3. [Priority clasee](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#priorityclass)
+
+
+### User Stories (Optional)
+
+#### Story 1 - Pod VS NetworkPolicy
+
+A user has pods which listen on the network and network policies which help protect those pods. 
+While namespace deletion, there could be cases that NetworkPolicy has deleted while the pods are running 
+which cause the security concern of having Pods running unprotected.
+
+After this feature was introduced, we would have NetworkPolicy in lower DeletionOrderPriority than the Pods to 
+avoid the above security concern.
+
+#### Story 2 - having finalizer conflicts with deletion order
+
+E.g. the NetworkPolicy deletion priority would be set lower than the pod and we could expect the Pod be deleted 
+before NetworkPolicy. However, if even one pod has a finalizer which is waiting for network policies (which is opaque to Kubernetes), 
+it will cause dependency loops and block the deletion process.
+
+Refer to the section `Handling Cyclic Dependencies`.
+
+#### Story 2 - having ownerReference conflicts with deletion order
+
+When deciding the deletion priority for resources, it should take ownerReference into consideration. 
+E.g. the deployment VS pod. However, it should not matter much in terms of namespace deletion. 
+Namespace deletion specifically uses `metav1.DeletePropagationBackground` and all resources would be deleted and the ownerReference 
+dependencies would be handled by the garbage collection.
```

Comment:
the two "Story 2"s aren't user stories. They're more like "Notes/Constraints/Caveats".